### PR TITLE
Add first version of synchronous IPC (#77) and test case

### PR DESCRIPTION
**Pull request by cstanfill**
sendrecv() logically takes a thread id to send to and some registers.
It sends the registers directly to the given thread and immediately
switches stacks to that thread.  Once that is done, the sending thread
becomes a receiving thread and doesn't wake up until it receives a
message back.

The send can only succeed if the thread provided is currently blocked
in another sendrecv() call.  In order to bootstrap this, a thread can
'send' a message to the kernel via thread id 0, which simply puts that
thread into receiving mode without sending anything (it becomes a
daemon).

If all the remaining threads are in receiving mode, they are assumed to
all be daemons and the system shuts down.  This is symmetric to the
bootstrapping behavior before - otherwise it would be impossible to shut
the system down after any thread called sendrecv() in daemon mode.

### CODE REVIEW

**Comment by cstanfill** [comment]
I updated this to include the 'generational cache' model of thread IDs we talked about. It messes with all the thread IDs printed in tests, unfortunately :\

File: kernel/ipc.c [hunk 1]
```diff
@@ -0,0 +1,40 @@
+#include "ipc.h"
+
+#include "com.h"
+#include "pagefault.h"
+#include "threads.h"
+#include "util.h"
+#include "memory-map.h"
+
+sendrecv_status sendrecv(sendrecv_op* usr_op) {
+  sendrecv_op op;
```
Comment (gsilvis) [inline kernel/ipc.c]: I think it's clearer if you specifically copy just the send-message here, and specifically copy the recv-message back at the end.

File: kernel/threads.h [hunk 2]
```diff
@@ -22,6 +23,12 @@ enum fpu_state {
   THREAD_FPU_STATE_ACTIVE,    /* there's an FPU buf allocated */
 };
 
+enum ipc_state {
+  IPC_NOT_RECEIVING,
+  IPC_RECEIVING,      /* currently waiting for an ipc message */
+  IPC_RECEIVED,       /* in the middle of receiving an ipc handoff */
```
Comment (gsilvis) [inline kernel/threads.h]: This state is a little sad.  With the current structure, where you drop down into 'schedule' before reaching the other thread, it's probably the best you can do, though.

File: kernel/threads.h [hunk 3]
```diff
@@ -38,6 +45,8 @@ typedef struct {
   size_t text_length;
   enum fpu_state fpu_state;
   char (*fpu_buf)[512];
+  enum ipc_state ipc_state;
+  ipc_msg inbox;  /* Only valid if ipc_state is RECEIVING */
```
Comment (gsilvis) [inline kernel/threads.h]: I think you mean 'RECEIVED'

File: kernel/threads.h [hunk 4]
```diff
@@ -59,6 +69,8 @@ void clone_thread (uint64_t fork_rsp);
 void spawn_within_vm_space (uint64_t rip, uint64_t rsp);
 void com_print_backtrace (void);
 
+tcb* get_tcb(uint64_t tid);
```
Comment (gsilvis) [inline kernel/threads.h]: super nit: I've been trying to write these C-style: tcb *get_tcb()

File: kernel/syscall-defs.h [hunk 5]
```diff
@@ -3,6 +3,23 @@
...
+  unsigned long r2;
+} ipc_msg;
+
+typedef struct {
+  ipc_msg send; /* Filled in by the user */
+  ipc_msg recv; /* Filled in by the kernel on a receive */
+} sendrecv_op;
+
+typedef enum {
+  SEND_FAILED = 0,
```
Comment (gsilvis) [inline kernel/syscall-defs.h]: Can you make 0 the successful case, so 'if (sendrecv())' works?

File: kernel/threads.c [hunk 6]
```diff
@@ -51,6 +51,7 @@ static tcb *create_thread (void *text, size_t length, vmcb *vm_space) {
       tcbs[i].state = TS_EXIST;
       tcbs[i].stack_top = &kernel_stack[0];
       tcbs[i].fpu_state = THREAD_FPU_STATE_INACTIVE;
+      tcbs[i].ipc_state = IPC_NOT_RECEIVING;
       total_threads++;
```
Comment (gsilvis) [inline kernel/threads.c]: Recalculating the number of tasks each time in 'choose_task' is fine, but can you delete 'total_threads' then?

(Alternatively, you could have the 'send to the kernel' case also reach in and decrement total_threads... but that might get hairy once you can also kill daemons.  So statelessly checking every time is pretty reasonable.)

Comment (cstanfill) [reply kernel/threads.c]: I didn't realize that was unused now - gladly deleted :)

File: kernel/syscall-defs.h (hunk 5 above)
Comment (cstanfill) [reply kernel/syscall-defs.h]: Sure

File: userland/test-ipc/main.c [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    op.send.r2 = 5;
+    switch (sendrecv(&op)) {
+      case SEND_FAILED:
+        DEBUG("SEND_FAILED");
+        return;
+      case RECEIVE_FAILED:
+        DEBUG("RECEIVE FAILED");
+        return;
+      case MESSAGE_RECEIVED:
+        if (op.recv.r1 + op.recv.r2 == i * 2 && op.recv.r1 - op.recv.r2 == 10) {
```
Comment (allanlw) [inline userland/test-ipc/main.c]: wtf is this math

**Pull request merged**